src/util/pup.h

```diff
@@ -208,85 +240,30 @@ class er {
-  void operator()(long double &v)     {(*this)(&v,1);}
-#endif
-  void operator()(bool &v)         {(*this)(&v,1);}
-#ifdef CMK_PUP_LONG_LONG
-  void operator()(CMK_PUP_LONG_LONG &v) {(*this)(&v,1);}
-  void operator()(unsigned CMK_PUP_LONG_LONG &v) {(*this)(&v,1);}
-#endif
-#if CMK_HAS_INT16
-  void operator()(CmiInt16 &v) {(*this)(&v,1);}
-  void operator()(CmiUInt16 &v) {(*this)(&v,1);}
-#endif
+  template<class T>
+  void operator()(T &v)               {(*this)(&v,1);}
+
   void operator()(void* &v,void* sig) {(*this)(&v,1,sig);}
-  
-//For arrays:
-  //Integral types:
-  void operator()(signed char *a,size_t nItems)
-    {bytes((void *)a,nItems,sizeof(signed char),Tchar);}
-#if CMK_SIGNEDCHAR_DIFF_CHAR
-  void operator()(char *a,size_t nItems)
-    {bytes((void *)a,nItems,sizeof(char),Tchar);}
-#endif
-  void operator()(short *a,size_t nItems)
-    {bytes((void *)a,nItems,sizeof(short),Tshort);}
-  void operator()(int *a,size_t nItems)
-    {bytes((void *)a,nItems,sizeof(int),Tint);}
-  void operator()(long *a,size_t nItems)
-    {bytes((void *)a,nItems,sizeof(long),Tlong);}
-
-  //Unsigned integral types:
-  void operator()(unsigned char *a,size_t nItems)
-    {bytes((void *)a,nItems,sizeof(unsigned char),Tuchar);}
-  void operator()(unsigned short *a,size_t nItems)
-    {bytes((void *)a,nItems,sizeof(unsigned short),Tushort);}
-  void operator()(unsigned int *a,size_t nItems)
-    {bytes((void *)a,nItems,sizeof(unsigned int),Tuint);}
-  void operator()(unsigned long *a,size_t nItems)
-    {bytes((void *)a,nItems,sizeof(unsigned long),Tulong);}
-
-  //Floating-point types:
-  void operator()(float *a,size_t nItems)
-    {bytes((void *)a,nItems,sizeof(float),Tfloat);}
-  void operator()(double *a,size_t nItems)
-    {bytes((void *)a,nItems,sizeof(double),Tdouble);}
 
-#if CMK_LONG_DOUBLE_DEFINED
-  void operator()(long double *a,size_t nItems)
-    {bytes((void *)a,nItems,sizeof(long double),Tlongdouble);}
-#endif
+  //For arrays:
+  template<class T>
+  void operator()(T *a,size_t nItems) {
+    bytes((void *)a,nItems, sizeof(T), getXlateDataType(a));
+  }
 
-  //For bools:
-  void operator()(bool *a,size_t nItems)
-    {bytes((void *)a,nItems,sizeof(bool),Tbool);}
+  template<class T>
+  void pup_buffer(T *&a, size_t nItems) {
+    if(isUnpacking()) a = (T *)malloc(nItems * sizeof(T));
+    bytes((void *)a,nItems, sizeof(T), getXlateDataType(a));
+    if(isPacking()) free(a);
+  }
 
-#ifdef CMK_PUP_LONG_LONG
-  void operator()(CMK_PUP_LONG_LONG *a,size_t nItems)
-    {bytes((void *)a,nItems,sizeof(CMK_PUP_LONG_LONG),Tlonglong);}
-  void operator()(unsigned CMK_PUP_LONG_LONG *a,size_t nItems)
-    {bytes((void *)a,nItems,sizeof(unsigned CMK_PUP_LONG_LONG),Tulonglong);}
-#endif
-#if CMK_HAS_INT16
-  void operator()(CmiInt16 *a,size_t nItems)
-    {bytes((void *)a,nItems,sizeof(CmiInt16),Tint128);}
-  void operator()(CmiUInt16 *a,size_t nItems)
-    {bytes((void *)a,nItems,sizeof(CmiUInt16),Tuint128);}
-#endif
+  template<class T>
```

Comment:
Maybe put a comment here that this is the same as above but with a custom allocator, deallocator